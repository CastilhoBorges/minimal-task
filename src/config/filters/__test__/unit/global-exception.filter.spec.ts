import { ArgumentsHost, HttpException, HttpStatus } from '@nestjs/common';
import { Response } from 'express';
import { HttpArgumentsHost } from '@nestjs/common/interfaces';
import { BaseException } from '../../../exceptions/base.exception';
import { GlobalExceptionFilter } from '../../global-exception.filter';

describe('GlobalExceptionFilter', () => {
  let filter: GlobalExceptionFilter;
  let mockResponse: Partial<Response>;
  let mockHttpArgumentsHost: Partial<HttpArgumentsHost>;
  let mockHost: Partial<ArgumentsHost>;

  beforeEach(() => {
    filter = new GlobalExceptionFilter();
    mockResponse = {
      status: jest.fn().mockReturnThis(),
      json: jest.fn(),
    };
    mockHttpArgumentsHost = {
      getResponse: jest.fn().mockReturnValue(mockResponse),
      getRequest: jest.fn(),
      getNext: jest.fn(),
    };
    mockHost = {
      switchToHttp: jest.fn(() => mockHttpArgumentsHost as HttpArgumentsHost),
    };
  });

  it('should handle a BaseException and match snapshot', () => {
    const exception = new BaseException({
      name: 'BusinessException',
      message: 'Business logic error',
      statusCode: HttpStatus.BAD_REQUEST,
      metadata: { code: 'INVALID_ACTION' },
      stack: 'stack_trace_here',
    });

    filter.catch(exception, mockHost as ArgumentsHost);

    expect(mockResponse.status).toHaveBeenCalledWith(HttpStatus.BAD_REQUEST);
    expect(mockResponse.json).toHaveBeenCalledWith(expect.any(Object));
    expect(mockResponse.json).toMatchSnapshot();
  });

  it('should handle an HttpException and match snapshot', () => {
    const exception = new HttpException(
      'Unauthorized access',
      HttpStatus.UNAUTHORIZED,
    );

    filter.catch(exception, mockHost as ArgumentsHost);

    expect(mockResponse.status).toHaveBeenCalledWith(HttpStatus.UNAUTHORIZED);
    expect(mockResponse.json).toHaveBeenCalledWith(expect.any(Object));
    expect(mockResponse.json).toMatchSnapshot();
  });

  it('should handle a generic Error and match snapshot', () => {
    const exception = new Error('Unexpected server failure');

    filter.catch(exception, mockHost as ArgumentsHost);

    expect(mockResponse.status).toHaveBeenCalledWith(
      HttpStatus.INTERNAL_SERVER_ERROR,
    );
    expect(mockResponse.json).toHaveBeenCalledWith(expect.any(Object));
    expect(mockResponse.json).toMatchSnapshot();
  });

  it('should handle an unknown exception and match snapshot', () => {
    const exception = { unknown: 'Unknown exception' };

    filter.catch(exception, mockHost as ArgumentsHost);

    expect(mockResponse.status).toHaveBeenCalledWith(
      HttpStatus.INTERNAL_SERVER_ERROR,
    );
    expect(mockResponse.json).toHaveBeenCalledWith(expect.any(Object));
    expect(mockResponse.json).toMatchSnapshot();
  });
});
