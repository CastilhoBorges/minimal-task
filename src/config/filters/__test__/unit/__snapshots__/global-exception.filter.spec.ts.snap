// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GlobalExceptionFilter should handle a BaseException and match snapshot 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "message": "Business logic error",
        "statusCode": 400,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GlobalExceptionFilter should handle a generic Error and match snapshot 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "message": "Unexpected server failure",
        "statusCode": 500,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GlobalExceptionFilter should handle an HttpException and match snapshot 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "message": "Unauthorized access",
        "statusCode": 401,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GlobalExceptionFilter should handle an unknown exception and match snapshot 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "message": "An unexpected error occurred",
        "statusCode": 500,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
